### PR TITLE
Add worker thread to ImageBuffer

### PR DESCRIPTION
**Pull request by MatthijsBurgh**
Merge together with https://github.com/tue-robotics/ed_navigation/pull/30, https://github.com/tue-robotics/ed_robocup/pull/24

### CODE REVIEW

File: ed_sensor_integration/src/kinect/image_buffer.cpp [hunk 1]
```diff
@@ -91,33 +97,50 @@ bool ImageBuffer::waitForRecentImage(const std::string& root_frame, rgbd::ImageC
...
+    sensor_pose = recent_image_.second;
+
+    recent_image_.first.reset(); // Invalidate the most recent image
+
+    return true;
+}
+
+// ----------------------------------------------------------------------------------------------------
+
+bool ImageBuffer::getMostRecentImageTF()
```
Comment (PetervDooren) [inline ed_sensor_integration/src/kinect/image_buffer.cpp:121]: It is currently missing the buffer functionality this thing was made for in the first place. If images and tf dont come in at the same time images will be discarded.

**Review by PetervDooren — CHANGES_REQUESTED** [review]

File: ed_sensor_integration/src/kinect/image_buffer.cpp [hunk 2]
```diff
@@ -158,9 +158,20 @@ bool ImageBuffer::getMostRecentImageTF()
                                 << ros::Time(rgbd_image->getTimestamp()));
                 ROS_WARN_STREAM("[IMAGE_BUFFER] Image too old to look-up tf: image timestamp = " << std::fixed
                                 << ros::Time(rgbd_image->getTimestamp()));
+                return false;
+            }
+            else
+            {
+                if(!tf_listener_->waitForTransform(root_frame_, rgbd_image->getFrameId(), ros::Time(rgbd_image->getTimestamp()), ros::Duration(0.5)))
```
Comment (PetervDooren) [inline ed_sensor_integration/src/kinect/image_buffer.cpp]: With this you cannot guarantee the execution rate of 20Hz(or whatever you initialised it to) of your worker thread. It also means you get the same problem if tf lags by 0.5seconds. 
And this is a magic number.

**Review by PetervDooren — CHANGES_REQUESTED** [review]

File: ed_sensor_integration/src/kinect/image_buffer.cpp (hunk 1 above)
Comment (MatthijsBurgh) [reply ed_sensor_integration/src/kinect/image_buffer.cpp:121]: Fixed by update using forward_list

File: ed_sensor_integration/src/kinect/image_buffer.cpp (hunk 2 above)
Comment (MatthijsBurgh) [reply ed_sensor_integration/src/kinect/image_buffer.cpp]: After update it tries to get TF for the most recent image. Only when TF is found for an image or an image is too old, images get dropped.

**Comment by MatthijsBurgh** [comment]
Behaviour:
```
HERO-FREE-MODE
[ WARN][/hero/ed][1648652348.808379974]: [IMAGE_BUFFER] Could not get sensor pose: "map" passed to lookupTransform argument target_frame does not exist. 
[ WARN][/hero/ed][1648652348.851867090]: [IMAGE_BUFFER] Could not get sensor pose: Could not find a connection between 'map' and 'head_rgbd_sensor_rgb_frame' because they are not part of the same tree.Tf has two or more unconnected trees.
[ WARN][/hero/ed][1648652348.877754358]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation 0.092289208s into the past.  Requested time 1648652348.703111649 but the earliest data is at time 1648652348.795400858, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ WARN][/hero/ed][1648652348.901090090]: [IMAGE_BUFFER] Could not get sensor pose: Could not find a connection between 'map' and 'head_rgbd_sensor_rgb_frame' because they are not part of the same tree.Tf has two or more unconnected trees.
[ WARN][/hero/ed][1648652348.907684501]: [IMAGE_BUFFER] Could not get sensor pose: Could not find a connection between 'map' and 'head_rgbd_sensor_rgb_frame' because they are not part of the same tree.Tf has two or more unconnected trees.
[ WARN][/hero/ed][1648652348.926749663]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation 0.058699607s into the past.  Requested time 1648652348.736701250 but the earliest data is at time 1648652348.795400858, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ WARN][/hero/ed][1648652348.950982765]: [IMAGE_BUFFER] Could not get sensor pose: Could not find a connection between 'map' and 'head_rgbd_sensor_rgb_frame' because they are not part of the same tree.Tf has two or more unconnected trees.
[ WARN][/hero/ed][1648652348.957642363]: [IMAGE_BUFFER] Could not get sensor pose: Could not find a connection between 'map' and 'head_rgbd_sensor_rgb_frame' because they are not part of the same tree.Tf has two or more unconnected trees.
[ WARN][/hero/ed][1648652348.976784752]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation 0.025183438s into the past.  Requested time 1648652348.770217419 but the earliest data is at time 1648652348.795400858, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ WARN][/hero/ed][1648652349.007666175]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation 0.147622869s into the past.  Requested time 1648652348.837249279 but the earliest data is at time 1648652348.984872103, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ WARN][/hero/ed][1648652349.026803490]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652348.837249279
[ WARN][/hero/ed][1648652349.051075578]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652348.870804787
[ WARN][/hero/ed][1648652349.057669978]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652348.870804787
[ WARN][/hero/ed][1648652349.076768569]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652348.904320717
[ WARN][/hero/ed][1648652349.100977536]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652348.904320717
[ WARN][/hero/ed][1648652349.107667176]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652348.904320717
[ WARN][/hero/ed][1648652349.150998745]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652348.937723637
[ WARN][/hero/ed][1648652349.157659525]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652348.937723637
[ WARN][/hero/ed][1648652349.176745355]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652348.971029043
[ WARN][/hero/ed][1648652349.200950516]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652348.971029043

EMERGENCY BUTTON PRESSED

[ERROR][/hero/ed][1648652421.226647359]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation -9.831385700s into the future.  Requested time 1648652421.063069344 but the latest data is at time 1648652411.231683731, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ WARN][/hero/ed][1648652421.226680775]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation -9.831385700s into the future.  Requested time 1648652421.063069344 but the latest data is at time 1648652411.231683731, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
.
.
.
[ WARN][/hero/ed][1648652427.676716457]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation -16.266101924s into the future.  Requested time 1648652427.497785568 but the latest data is at time 1648652411.231683731, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ WARN][/hero/ed][1648652427.700940882]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation -16.266101924s into the future.  Requested time 1648652427.497785568 but the latest data is at time 1648652411.231683731, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]

EMERGENCY BUTTON RELEASED

[ WARN][/hero/ed][1648652427.757606774]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation 0.161607731s into the past.  Requested time 1648652427.564671278 but the earliest data is at time 1648652427.726279020, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ WARN][/hero/ed][1648652427.776739739]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation 0.161607731s into the past.  Requested time 1648652427.564671278 but the earliest data is at time 1648652427.726279020, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ WARN][/hero/ed][1648652427.801006522]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation 0.161607731s into the past.  Requested time 1648652427.564671278 but the earliest data is at time 1648652427.726279020, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ WARN][/hero/ed][1648652427.826735559]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation 0.094625462s into the past.  Requested time 1648652427.631653547 but the earliest data is at time 1648652427.726279020, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ WARN][/hero/ed][1648652427.850954450]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation 0.094625462s into the past.  Requested time 1648652427.631653547 but the earliest data is at time 1648652427.726279020, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ WARN][/hero/ed][1648652427.857622141]: [IMAGE_BUFFER] Could not get latest sensor pose (probably because tf is still initializing): Lookup would require extrapolation 0.027674425s into the past.  Requested time 1648652427.698604584 but the earliest data is at time 1648652427.726279020, when looking up transform from frame [head_rgbd_sensor_rgb_frame] to frame [map]
[ERROR][/hero/ed][1648652427.876804493]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652427.698604584
[ WARN][/hero/ed][1648652427.876841407]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652427.698604584
[ WARN][/hero/ed][1648652427.900959782]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652427.698604584
[ WARN][/hero/ed][1648652427.926701895]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652427.765636683
[ WARN][/hero/ed][1648652427.950935457]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652427.765636683
[ WARN][/hero/ed][1648652427.957612015]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652427.799153566
[ WARN][/hero/ed][1648652427.976747245]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652427.799153566
[ WARN][/hero/ed][1648652428.000984528]: [IMAGE_BUFFER] Image too old to look-up tf: image timestamp = 1648652427.799153566
```

**Review by PetervDooren — APPROVED** [review]
Looks good to me. Do consider whether we need all these ros errors or warnings.

**Pull request merged**